Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -444,10 +451,61 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
   }
 }
 
+public protocol Mergable 
+where Self: Differentiable {
+  @differentiable
+  static func +(lhs: Self, rhs: Self) -> Self
+}
+
+public enum MergeMode {
+  case sum
+}
+
+public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer 
+where Cell.TimeStepOutput: Mergable {
+  public typealias Input = [Cell.TimeStepInput]
+  public typealias Output = [Cell.TimeStepOutput]
+  public typealias Merge = @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput
+
+  @noDerivative public let mergeMode: MergeMode
+  public var forward, backward: RecurrentLayer<Cell>
+
+  public init(_ cell: @autoclosure () -> Cell, mergeMode: MergeMode = .sum) {
+    forward = RecurrentLayer(cell())
+    backward = RecurrentLayer(cell())
+    self.mergeMode = mergeMode
+  }
+
+  @differentiable
+  public func callAsFunction(_ inputs: Input) -> Output {
+    let inputsReversed = withoutDerivative(at: Array(inputs.reversed()))
+
+    let forwardOutputs = forward(inputs)
+    let backwardOutputs = backward(inputsReversed)
+
+    var outputs = Output()
+
+    for i in  0 ..< withoutDerivative(at: inputs.count) {
+      switch mergeMode {
+      case .sum:
+        outputs.append(forwardOutputs[i] + backwardOutputs[i])
+      }
+    }
+
+    return outputs
+  }
+
+  @differentiable
+  public func lastOutput(_ inputs: Input) -> Cell.TimeStepOutput {
+    self(inputs)[withoutDerivative(at: inputs.count - 1)]
+  }
+}
+
 extension RecurrentLayer: Equatable where Cell: Equatable {}
 extension RecurrentLayer: AdditiveArithmetic where Cell: AdditiveArithmetic {}
 
 public typealias BasicRNN<Scalar: TensorFlowFloatingPoint> = RecurrentLayer<BasicRNNCell<Scalar>>
+public typealias BiBasicRNN<Scalar: TensorFlowFloatingPoint> = BidirectionalRecurrentLayer<BasicRNNCell<Scalar>>
```

Comment:
The Swift API Design Guidelines emphasize clear names rather than short ones: [clarity over brevity](https://swift.org/documentation/api-design-guidelines/#clarity-over-brevity).

How about naming typealiases as `BidirectionalXXX`? I think it's very clear and reasonably short.
```swift
public typealias BidirectionalBasicRNN<Scalar: TensorFlowFloatingPoint> = BidirectionalRecurrentLayer<BasicRNNCell<Scalar>>
public typealias BidirectionalLSTM<Scalar: TensorFlowFloatingPoint> = BidirectionalRecurrentLayer<LSTMCell<Scalar>>
public typealias BidirectionalGRU<Scalar: TensorFlowFloatingPoint> = BidirectionalRecurrentLayer<GRUCell <Scalar>>
```

Reply:
Yes, renamed.